### PR TITLE
Chip reliable message protocol

### LINKED ISSUE #3197: Define CRMP interface
#### Problem
We plan to implement CRMP as an separate Class in messaging layer, Apple would like to get agreed on the interface first.  

 #### Proposed Solution
Define CRMP interface for Messaging layer

### PR DESCRIPTION
**Pull request by kghost**
**Problem**
Implement CRMP as an separate Class in messaging layer

**Summary of Changes**
Define and Implementation of CRMP for messaging layer

Fixes #3197

### CODE REVIEW

File: src/messaging/CHIPReliableMessageManager.h
Comment (kghost) [inline src/messaging/CHIPReliableMessageManager.h:97]: This is a placeholder function, I will complete the implementation once exchange manager is added

File: src/messaging/CHIPMessageLayerUtils.cpp
Comment (woody-apple) [inline src/messaging/CHIPMessageLayerUtils.cpp:57]: We should make sure this works on Darwin too.... (dispatch based stuff)

Comment (kghost) [reply src/messaging/CHIPMessageLayerUtils.cpp:57]: I doesn't have Darwin build env, and I'm not sure how to check it. I'll try add CHIP_SYSTEM_CONFIG_USE_NETWORK_FRAMEWORK, so buildbot can check it for me

File: src/messaging/CHIPMessageLayerUtils.h
Comment (rwalker-apple) [inline src/messaging/CHIPMessageLayerUtils.h:32]: classes inside the chip namespace need not have Chip in their name, they're already named chip::Something

please add a `Messaging` namespace for these classes, too

File: src/messaging/CHIPReliableMessageConfig.h
Comment (rwalker-apple) [inline src/messaging/CHIPReliableMessageConfig.h:26]: our current convention is to use `#pragma once` instead of include guards, can you update these, please?

File: src/messaging/CHIPReliableMessageConfig.h
Comment (rwalker-apple) [inline src/messaging/CHIPReliableMessageConfig.h:31]: what happens if we leave clang-format on for these?

File: src/messaging/CHIPReliableMessageContext.cpp
Comment (rwalker-apple) [inline src/messaging/CHIPReliableMessageContext.cpp:29]: please place the module header as the first include

File: src/messaging/CHIPReliableMessageContext.cpp
Comment (rwalker-apple) [inline src/messaging/CHIPReliableMessageContext.cpp:200]: can we leave off the "RMP" on methods that the ReliableMessageContext implements to aid readability?

File: src/messaging/CHIPReliableMessageContext.h
Comment (rwalker-apple) [inline src/messaging/CHIPReliableMessageContext.h:33]: config headers should be first includes from another header

File: src/messaging/CHIPReliableMessageContext.h
Comment (rwalker-apple) [inline src/messaging/CHIPReliableMessageContext.h:24]: #pragma once, please, across all these headers?

**Review by rwalker-apple — DISMISSED** [review]

File: src/messaging/CHIPMessageLayerUtils.cpp
Comment (kghost) [reply src/messaging/CHIPMessageLayerUtils.cpp]: done

File: src/messaging/CHIPMessageLayerPrivate.h
Comment (kghost) [reply src/messaging/CHIPMessageLayerPrivate.h]: done

File: src/messaging/CHIPMessageLayerUtils.h
Comment (kghost) [reply src/messaging/CHIPMessageLayerUtils.h]: done

File: src/messaging/CHIPReliableMessageConfig.h
Comment (kghost) [reply src/messaging/CHIPReliableMessageConfig.h]: done

File: src/messaging/CHIPReliableMessageConfig.h
Comment (kghost) [reply src/messaging/CHIPReliableMessageConfig.h]: done

File: src/messaging/CHIPReliableMessageContext.cpp
Comment (kghost) [reply src/messaging/CHIPReliableMessageContext.cpp]: done

File: src/messaging/CHIPReliableMessageContext.cpp
Comment (kghost) [reply src/messaging/CHIPReliableMessageContext.cpp]: done

File: src/messaging/CHIPReliableMessageContext.h
Comment (kghost) [reply src/messaging/CHIPReliableMessageContext.h]: done

File: src/messaging/CHIPReliableMessageContext.h
Comment (kghost) [reply src/messaging/CHIPReliableMessageContext.h]: done

File: src/messaging/ReliableMessageContext.cpp
Comment (andy31415) [inline src/messaging/ReliableMessageContext.cpp:57]: nit: I would note that we deprecated GetFlag/SetFlag/ClearFlag in favor of BitFlags (typesafe version).

This would imply a bit of a refactor: flags would become enum values, members change a bit.

File: src/messaging/ReliableMessageContext.h
Comment (andy31415) [inline src/messaging/ReliableMessageContext.h:113]: Could we make data members be private with getters/setters as applicable?

File: src/messaging/ReliableMessageManager.cpp
Comment (andy31415) [inline src/messaging/ReliableMessageManager.cpp:338]: I assume the cast down-casts some larger value (32 or 64 bit?) to 16. Could we have a comment on why this is safe? I imagine tick counters to be potentially large values.

File: src/messaging/ReliableMessageManager.cpp
Comment (andy31415) [inline src/messaging/ReliableMessageManager.cpp:404]: continue here does not seem needed.

File: src/messaging/ReliableMessageManager.cpp
Comment (andy31415) [inline src/messaging/ReliableMessageManager.cpp:400]: we could just 'return true' here.

File: src/messaging/ReliableMessageManager.cpp
Comment (andy31415) [inline src/messaging/ReliableMessageManager.cpp:443]: could we make len be uint16_t instead and not need to cast? It seems as if Get and Set should support he same data size.

File: src/messaging/ReliableMessageManager.cpp
Comment (andy31415) [inline src/messaging/ReliableMessageManager.cpp:460]: Should we create a separate category of ReliableMessaging instead of ExchangeManager since this is now removed from ExchangeManager?

File: src/messaging/ReliableMessageManager.h
Comment (andy31415) [inline src/messaging/ReliableMessageManager.h:53]: This should be a structu (all public data members). If a class, data members should be private with getters/setters, however that seems overkill.

File: src/messaging/ReliableMessageManager.h
Comment (andy31415) [inline src/messaging/ReliableMessageManager.h:93]: data members should be private.

Unsure if doccomment is required ('ReliableMessageProtocolConfig' is obviously a 'ReliableMessageProtocol configuration').

**Review by andy31415 — COMMENTED** [review]
We need unit tests for such a complex functionality.

**Review by andy31415 — CHANGES_REQUESTED** [review]
Marking as 'request changes' for unit tests. All my other comments were nits.

File: src/messaging/Flags.h
Comment (kghost) [reply src/messaging/Flags.h:37]: These flags is for message layer (exchange context) internal use. I'll merge it with `exchange` PRs when they are ready

File: src/messaging/ReliableMessageContext.h
Comment (kghost) [reply src/messaging/ReliableMessageContext.h:48]: Done

File: src/messaging/ReliableMessageManager.cpp
Comment (kghost) [reply src/messaging/ReliableMessageManager.cpp:443]: done

File: src/messaging/ReliableMessageContext.cpp
Comment (kghost) [reply src/messaging/ReliableMessageContext.cpp:57]: done

File: src/messaging/ErrorCategory.h
Comment (andy31415) [inline src/messaging/ErrorCategory.h:20]: Defines error filtering and selection methods (no more utility class)

File: src/messaging/Flags.h
Comment (andy31415) [inline src/messaging/Flags.h:21]: defines flags for message reliable message processing?

File: src/messaging/ErrorCategory.h
Comment (kghost) [reply src/messaging/ErrorCategory.h]: done

File: src/messaging/Flags.h
Comment (kghost) [reply src/messaging/Flags.h]: done

File: src/messaging/ReliableMessageContext.cpp
Comment (andy31415) [inline src/messaging/ReliableMessageContext.cpp:92]: nit: I think BitFlags has a Set(flag, bool) member, so the functions below could be like:

`mFlags.Set(Flags::kFlagMsgRcvdFromPeer, inMsgRecvdFromPeer);`

File: src/messaging/ReliableMessageContext.cpp
Comment (andy31415) [inline src/messaging/ReliableMessageContext.cpp:238]: could we avoid singletons and instead make the message manager pointer a member of this class?

This comment applies throughout this file: ReliableMessageManager seems too coupled and used everywhere.

File: src/messaging/ReliableMessageManager.cpp
Comment (andy31415) [inline src/messaging/ReliableMessageManager.cpp:149]: how about inverting the conditions to decrease the indent level? Something like:

```
for (int i = 0; i < CHIP_CONFIG_RMP_RETRANS_TABLE_SIZE; i++)
{
￼   ReliableMessageContext * rc = RetransTable[i].rc;
￼   if (!rc) {
       continue;
    }

    if (0 == RetransTable[i].nextRetransTimeTick) {
       continue;
    }

    .....
```

File: src/messaging/tests/TestReliableMessageProtocol.cpp
Comment (andy31415) [inline src/messaging/tests/TestReliableMessageProtocol.cpp:50]: Generally tests should never sleep.

I believe RMP uses timers - could we provide a timer implementation to the functions, so that tests have a 'fake timer' that can be controller on when to fire by the tests themselves?

File: src/messaging/tests/TestReliableMessageProtocol.cpp
Comment (andy31415) [inline src/messaging/tests/TestReliableMessageProtocol.cpp:244]: We should try to remove this kind of coupling.

**Review by andy31415 — APPROVED** [review]
I believe the current classes are a bit too coupled, however I don't want to hold off the PR too much either, hence approving.

I imagine we  want to iterate a bit over it, specifically to remove hardcoded dependencies between classes (context calling manager directly and dependency on a system-provided timer).

File: src/messaging/ReliableMessageContext.cpp
Comment (kghost) [reply src/messaging/ReliableMessageContext.cpp:92]: done

File: src/messaging/ReliableMessageContext.cpp
Comment (kghost) [reply src/messaging/ReliableMessageContext.cpp:238]: done

File: src/messaging/ReliableMessageManager.cpp
Comment (kghost) [reply src/messaging/ReliableMessageManager.cpp:149]: done

File: src/messaging/tests/TestReliableMessageProtocol.cpp
Comment (kghost) [reply src/messaging/tests/TestReliableMessageProtocol.cpp:244]: done

File: src/messaging/tests/TestReliableMessageProtocol.cpp
Comment (kghost) [reply src/messaging/tests/TestReliableMessageProtocol.cpp:50]: It is not timer which is causing problem. The event-loop in system layer is not running, and I'm triggering the timer manually.

The problem is that there is no way to mock `System::Timer::GetCurrentEpoch()` function. I tried to make it not using the function, but as a result, all member functions of the manager class and the context class need to take an extra arguments `now`, I think it is an overshoot.

Currently the sleep time is limited to several milliseconds, and the test cases are running consistently w/o any problem. before we introduce a way to mock the function I think currently it is fine.

File: src/messaging/ReliableMessageContext.cpp
Comment (andy31415) [inline src/messaging/ReliableMessageContext.cpp:105]: Simple set can be used here and in SetPeerRequestedAck below as well.

Comment (andy31415) [reply src/messaging/ReliableMessageContext.cpp:105]: In the DropAck method too.

File: src/messaging/tests/TestReliableMessageProtocol.cpp
Comment (rwalker-apple) [inline src/messaging/tests/TestReliableMessageProtocol.cpp:21]: comment is copy-pasta'd?

File: src/messaging/tests/TestReliableMessageProtocolDriver.cpp
Comment (rwalker-apple) [inline src/messaging/tests/TestReliableMessageProtocolDriver.cpp:21]: more copy-paste?

**Review by rwalker-apple — COMMENTED** [review]
how will this interact with ExchangeMgr (if at all)?

File: src/messaging/tests/TestReliableMessageProtocol.cpp
Comment (rwalker-apple) [inline src/messaging/tests/TestReliableMessageProtocol.cpp:68]: do we need tests for these firing at the right times?

**Review by rwalker-apple — APPROVED** [review]

File: src/messaging/ReliableMessageManager.h
Comment (yufengwangca) [inline src/messaging/ReliableMessageManager.h:104]: Where is the implementation of SendMessage function?

Comment (kghost) [reply src/messaging/ReliableMessageManager.h:104]: It will be integrate with `ExchangeContext`. We will find a way to associate `ReliableMessageContext` with `ExchangeContext` and here `SendMessage` will probably propagate to `SendMessage` in `ExchangeManager`.

Comment (yufengwangca) [reply src/messaging/ReliableMessageManager.h:104]: Thanks, I think this need to be implemented with SessionManager, just like what ExchangeContext does today. After CRMP is merged, ExchangeContext will switch to use CRMP to send message instead of using SessionManager.

**Review by yufengwangca — APPROVED** [review]

File: src/messaging/ReliableMessageManager.h
Comment (yufengwangca) [reply src/messaging/ReliableMessageManager.h:104]: I think we can leave this implementation to another PR. Basically, CRMP is an internal module of messaging layer and ExchangeManager /ExchangeContext is its client. For TCP/BLE message, ExchangeContext use SecureSessionManager to send the message directly, for UDP message, ExchangeContext will use CRMP to send the message instead.

**Review by jelderton — APPROVED** [review]

**Comment by rwalker-apple** [comment]
@kghost are you still working on this?

**Comment by bzbarsky-apple** [comment]
@kghost There's non-compiling `DEBUG` code in this PR that breaks iOS builds.  When you get a chance, can you look at https://github.com/project-chip/connectedhomeip/issues/3542 please?

**Pull request merged**
